### PR TITLE
Prevent unnecessary activation of the Python extension

### PR DESCRIPTION
**Pull request by DonJayamanne**
For #5193

### CODE REVIEW

File: src/client/api/pythonApi.ts [hunk 1]
```diff
@@ -49,13 +49,32 @@ import {
...
 
     constructor(
         @inject(IExtensions) private readonly extensions: IExtensions,
+        @inject(IDisposableRegistry) private readonly disposables: IDisposableRegistry,
         @inject(IPythonExtensionChecker) private extensionChecker: IPythonExtensionChecker
-    ) {}
+    ) {
+        const previouslyInstalled = this.extensionChecker.isPythonExtensionInstalled;
+        if (!previouslyInstalled) {
+            this.extensions.onDidChange(
```
Comment (DonJayamanne) [inline src/client/api/pythonApi.ts:67]: If the extension is installed, then register the hooks

File: src/client/datascience/kernel-launcher/localKernelFinder.ts [hunk 2]
```diff
@@ -90,12 +91,13 @@ export class LocalKernelFinder implements ILocalKernelFinder {
...
             // Get list of all of the specs
             const kernels = await this.listKernels(resource, cancelToken);
+            const isPythonNb = isPythonNotebook(option);
 
             // Always include the interpreter in the search if we can
             const interpreter =
                 option && isInterpreter(option)
                     ? option
-                    : resource && this.extensionChecker.isPythonExtensionInstalled
+                    : resource && isPythonNb && this.extensionChecker.isPythonExtensionInstalled
```
Comment (DonJayamanne) [inline src/client/datascience/kernel-launcher/localKernelFinder.ts:100]: Use active interpreter only if we know this is a python notebook

File: src/client/datascience/notebookStorage/nativeEditorStorage.ts [hunk 3]
```diff
@@ -220,11 +216,6 @@ export class NativeEditorStorage implements INotebookStorage {
             // eslint-disable-next-line @typescript-eslint/no-explicit-any
             return (notebookData.metadata.language_info.codemirror_mode as any).version;
         }
-        // Use the active interpreter if allowed
-        if (this.extensionChecker.isPythonExtensionInstalled) {
```
Comment (DonJayamanne) [inline src/client/datascience/notebookStorage/nativeEditorStorage.ts:224]: This is totally unnecessary, we're just storing this in metadata for language, today we generate all of this information from the kernel, hence we have more accurate information than this.

Comment (DonJayamanne) [reply src/client/datascience/notebookStorage/nativeEditorStorage.ts:224]: Also this was more stuff that slowed down notebooks, unnecessary when creating blank notebooks.

File: src/client/api/pythonApi.ts [hunk 4]
```diff
@@ -291,16 +325,17 @@ export class InterpreterService implements IInterpreterService {
...
-                .getApi()
-                .then((api) => {
-                    if (!this.eventHandlerAdded) {
-                        this.eventHandlerAdded = true;
-                        api.onDidChangeInterpreter(() => this.didChangeInterpreter.fire(), this, this.disposables);
-                    }
-                })
-                .catch(noop);
+        if (this.extensionChecker.isPythonExtensionInstalled) {
+            if (this.extensionChecker.isPythonExtensionActive && this.eventHandlerAdded) {
```
Comment (IanMatthewHuff) [inline src/client/api/pythonApi.ts:329]: Why is the check for this.eventHandlerAdded here? If eventHandlerAdded === true then hookup is a noop.

File: src/client/api/types.ts [hunk 5]
```diff
@@ -19,12 +19,14 @@ export interface ILanguageServer extends Disposable {
 
 export const IPythonApiProvider = Symbol('IPythonApi');
 export interface IPythonApiProvider {
+    onDidActivatePythonExtension: Event<void>;
```
Comment (IanMatthewHuff) [inline src/client/api/types.ts:22]: Not fully sure if I'm suggesting this. But does this seem to you like it should be on IPythonExtensionChecker? That has everything relevant to checking if the extension is installed or active.

**Review by IanMatthewHuff — CHANGES_REQUESTED** [review]
Looks solid. Just had one question on the interfaces and one spot that looked like maybe a bug with the event hookup.

File: src/client/api/pythonApi.ts [hunk 6]
```diff
@@ -49,13 +49,32 @@ import {
 @injectable()
 export class PythonApiProvider implements IPythonApiProvider {
     private readonly api = createDeferred<PythonApi>();
+    private readonly didActivatePython = new EventEmitter<void>();
```
Comment (joyceerhl) [inline src/client/api/pythonApi.ts:52]: Do we need to make sure we dispose of this object?

File: src/client/datascience/kernel-launcher/localKernelFinder.ts (hunk 2 above)
Comment (rchiodo) [reply src/client/datascience/kernel-launcher/localKernelFinder.ts:100]: Not sure this works with the interactive window

**Review by rchiodo — CHANGES_REQUESTED** [review]
Waiting on check that active interpreter is still used for interactive window

File: src/client/api/types.ts (hunk 5 above)
Comment (DonJayamanne) [reply src/client/api/types.ts:22]: No can do, its the PythonApiProvider that activates the Python extension. So only PythonAPIProvider knows if the python extension was activated or not.
VS Code doesn't fire events when extensions load.
Here we're only interested in situations when we load the python extension & other parts of the code can then react to that.

File: src/client/api/pythonApi.ts (hunk 6 above)
Comment (DonJayamanne) [reply src/client/api/pythonApi.ts:52]: Fixed.

File: src/client/api/pythonApi.ts (hunk 4 above)
Comment (DonJayamanne) [reply src/client/api/pythonApi.ts]: Fixed

**Review by DonJayamanne — COMMENTED** [review]
> Waiting on check that active interpreter is still used for interactive window

Don't quite follow this comment.

**Comment by rchiodo** [comment]
See the comment in localKernelFinder.ts

**Review by DavidKutu — APPROVED** [review]

File: src/client/datascience/kernel-launcher/localKernelFinder.ts (hunk 2 above)
Comment (DonJayamanne) [reply src/client/datascience/kernel-launcher/localKernelFinder.ts:100]: Aaah, i think i can use the resource to check that, thanks

**Review by rchiodo — APPROVED** [review]

File: src/client/datascience/kernel-launcher/localKernelFinder.ts (hunk 2 above)
Comment (DonJayamanne) [reply src/client/datascience/kernel-launcher/localKernelFinder.ts]: Fixed.

**Comment by DonJayamanne** [comment]
> See the comment in localKernelFinder.ts

Somehow that comment was missing when i replied. anyways.

**Review by IanMatthewHuff — APPROVED** [review]

**Pull request merged**